### PR TITLE
Speculative Decoding Example

### PR DESCRIPTION
**Pull request by marius-baseten**
This is a preliminary working example of draft-target-model based speculative decoding.

It depends on two TRT-LLM engines, where the target model needs the `max_draft_len` build arg set to support speculative decoding (see included build scripts).

There is still some work todo documented in comments and notes.

### CODE REVIEW

File: speculative_decoding/model/model.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                self._config["model_metadata"]["speculative_decoding"][
+                    "draft_tokenizer_repository"
+                ]
+            ),
+        )
+
+    async def predict(self, model_input) -> str | AsyncGenerator[str, None]:
+        request_id = str(os.getpid()) + str(next(self._request_id_counter))
+        request = helpers.GenerationRequest.parse_obj(model_input)
+        max_num_draft_tokens: int = self._config["model_metadata"][
```
Comment (aspctu) [inline speculative_decoding/model/model.py:103]: Would be good to move these out from model_metadata and into the build key. Also worth using Pydantic for this so you don't need to string index.

File: speculative_decoding/config.yaml [hunk 2]
```diff
@@ -1,28 +1,34 @@
...
 requirements:
-  - colorama
-  - csv
-  - huggingface_hub
-  - numpy
-  - pandas
-  - pydantic
-  - transformers
-  - tritonclient[all]
+- colorama
```
Comment (pankajroark) [inline speculative_decoding/config.yaml:19]: Good to fix versions here for predictable builds.

File: speculative_decoding/config.yaml [hunk 3]
```diff
@@ -1,28 +1,34 @@
...
+- huggingface_hub
+- numpy
+- pandas
+- pydantic
+- transformers
+- tritonclient[all]
 resources:
   accelerator: A100
+  cpu: '1'
```
Comment (pankajroark) [inline speculative_decoding/config.yaml:28]: Maybe remove these. A100 automatically comes with good defaults, these low numbers may be problematic and right sizing is not straightforward.

File: speculative_decoding/config.yaml [hunk 4]
```diff
@@ -1,28 +1,34 @@
...
+  memory: 2Gi
   use_gpu: true
 runtime:
   predict_concurrency: 256
+secrets: {}
```
Comment (pankajroark) [inline speculative_decoding/config.yaml:33]: We can remove this as it's the default.

File: speculative_decoding/engine_building/build_mistral.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import subprocess
+
+from huggingface_hub import snapshot_download
+
+# MODEL_NAME = "mistral-7B"
+# MODEL_HF_ID = "mistralai/Mistral-7B-v0.1"
+# Mixtral gets killed...
+MODEL_NAME = "mixtral-7B"
+MODEL_HF_ID = "mistralai/Mixtral-8x7B-v0.1"
```
Comment (pankajroark) [inline speculative_decoding/engine_building/build_mistral.py:9]: I think this should be https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2. @aspctu could you confirm?

**Review by pankajroark — COMMENTED** [review]
About half done, I'll review the rest tonight.

File: speculative_decoding/engine_building/build_gpt2.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+BUILD_PYTOHON_BIN = "/usr/bin/python"
+
+
+HF_DIR = f"/root/workbench/{MODEL_NAME}_hf"
+FT_DIR = f"/root/workbench/{MODEL_NAME}_ft"
+ENGINE_DIR = f"/root/workbench/{MODEL_NAME}_engine"
+
+MAX_DRAFT_LEN = 4
+
+# snapshot_download(
```
Comment (pankajroark) [inline speculative_decoding/engine_building/build_gpt2.py:17]: Remove commented code?

File: speculative_decoding/engine_building/build_mistral.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import subprocess
+
+from huggingface_hub import snapshot_download
+
+# MODEL_NAME = "mistral-7B"
+# MODEL_HF_ID = "mistralai/Mistral-7B-v0.1"
+# Mixtral gets killed...
```
Comment (pankajroark) [inline speculative_decoding/engine_building/build_mistral.py:7]: Remove commented code?

File: speculative_decoding/packages/spec_dec.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+from typing import Sequence
+
+import colorama
+import helpers  # From packages.
+import numpy as np
+import transformers
+import tritonclient.grpc.aio as triton_grpc
+
+
+class SpeculationState:
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:34]: Doc comment will be useful here, like what do we keep track of here, whether it is per request etc.

File: speculative_decoding/packages/spec_dec.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+    max_num_draft_tokens: int,
+    request_id: str,
+    result_queue: asyncio.Queue[str | QUEUE_SENTINEL] | None = None,
+    verbose: bool = False,
+    iteration_delay: float = 0.0,
+) -> SpeculationState:
+    """Runs the speculative decoding control flow loop betweeen draft and target model
+
+    * Model invocations are async inference requests to the triton server.
+    * For streaming, intermediate results are published in an optiona `result_queue`,
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:299]: Type: optiona -> optional

File: speculative_decoding/packages/spec_dec.py [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+       note that currently each result includes the context prefix, not just generated
+       text.
+    * For the loop to advance, it is necessary to await the results all inferernce
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:302]: typo: results `of`

File: speculative_decoding/packages/spec_dec.py [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+      requests. But if this coroutine function is scheduled as a task, e.g.
+      by `asyncio.ensure_future`, one python server can run multiple inferences
+      concurrently. The inference requests coming from those might can be collated to
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:305]: typo: `might`` seems left by mistake.

File: speculative_decoding/packages/spec_dec.py [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+            request.sampling_config,
+            request.bad_word_list,
+            request.stop_words_list,
+        )
+        state.update_draft(combined_draft)
+        if iteration_delay:
+            time.sleep(iteration_delay)
+
+        verification_inputs = state.get_verification_inputs()
+        if verification_inputs is None:
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:330]: Worth commenting when this case can be hit. It's perhaps when the draft token does not generate any new tokens in the code above in the while loop. It's not immediately clear when reading the code as to what conditions that can occur in and will be helpful to have as comment here.

File: speculative_decoding/packages/spec_dec.py [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+        with helpers.timeit(f"Generate({self._model_name}) - detokenize", skip=True):
+            output_text = self._tokenizer.decode(output_ids, skip_special_tokens=True)
+        return output_text, output_ids
+
+
+QUEUE_SENTINEL = None
+
+
+async def run_speculative_inference(
```
Comment (pankajroark) [inline speculative_decoding/packages/spec_dec.py:286]: Will be really good to have unit tests for this, the code is quite complex. We don't have a standard way of specifying unit tests in a truss right now, but for now we can make assumptions such as use of pytest, writes tests in a top level tests folder and a helper script to run those tests.

**Review by pankajroark — APPROVED** [review]
lgtm, minor comments.

File: speculative_decoding/packages/spec_dec.py (hunk 13 above)
Comment (marius-baseten) [reply speculative_decoding/packages/spec_dec.py:286]: Added to TODOs at top of file.

**Pull request merged**